cmd/ndm_daemonset/controller/devicestore.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2019 The OpenEBS Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+
+	"github.com/golang/glog"
+	apis "github.com/openebs/node-disk-manager/pkg/apis/openebs/v1alpha1"
+	"github.com/openebs/node-disk-manager/pkg/util"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// CreateDevice creates the Device resource in etcd
+func (c *Controller) CreateDevice(dvr apis.Device) {
+
+	dvrCopy := dvr.DeepCopy()
+	err := c.Clientset.Create(context.TODO(), dvrCopy)
+	if err == nil {
+		glog.Info("Created device object in etcd : ",
+			dvrCopy.ObjectMeta.Name)
+		return
+	}
+	glog.Infof("Dvr:%#v", dvrCopy)
+	glog.Error("Creation of device object failed : ", err)
+
+	/*
+	 * Creation may fail because resource is already exist in etcd.
+	 * This is possible when disk has been moved from one node to
+	 * another in a cluster and so device object need to be updated
+	 * with new owner i.e current node.
+	 */
+	err = c.UpdateDevice(dvr, nil)
+	if err == nil {
+		return
+	}
+
+	/*
+	 * Updation failure can be due to the fact that old node may have set
+	 * the status to Inactive after updateDvr has done the Get call, as
+	 * resource version will change with each update, so we have to retry.
+	 * Also if other node try to update resource version after updation is
+	 * successful here, the update call from that node will fail.
+	 */
+	glog.Info("Device status updated by other node, ",
+		"changing the ownership to this node : ", err)
+	err = c.UpdateDevice(dvr, nil)
```

Comment:
If the logic is to definitely update irrespective of resource version or simultaneous updates by other controllers/node, then probably Patch is a better option. It is definitive & will be quicker versus retrying & waiting for a successful update.

Reply:
We have seen cases, where none of the controllers (here node) will ever be able to update due to simultaneous updates. This leads to cases where the resource is never updated.

Reply:
If I look into Client's interfaces, it provide these three "Writer" interfaces :- a) Create b) Update and c) Delete.  I am sure you are talking about using some generic "patch" interface. Will sync-up tomorrow or whenever you are in office to understand more of it.